In [85]:
import pandas as pd 

In [86]:
df= pd.read_csv('C:/01_Kuliah/01K_Lainnya/CSV/buddymove_holidayiq.CSV')

In [87]:
display(df.head())
display(df.tail())

User Id  Sports  Religious  Nature  Theatre  Shopping  Picnic
0  User 1       2         77      79       69        68      95
1  User 2       2         62      76       76        69      68
2  User 3       2         50      97       87        50      75
3  User 4       2         68      77       95        76      61
4  User 5       2         98      54       59        95      86

User Id  Sports  Religious  Nature  Theatre  Shopping  Picnic
244  User 245      18        139     148      129       129     168
245  User 246      22        114     228      104        84     168
246  User 247      20        124     178      104       158     174
247  User 248      20        133     149      139       144     213
248  User 249      20        143     149      139       159     143

In [88]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249 entries, 0 to 248
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   User Id    249 non-null    object
 1   Sports     249 non-null    int64 
 2   Religious  249 non-null    int64 
 3   Nature     249 non-null    int64 
 4   Theatre    249 non-null    int64 
 5   Shopping   249 non-null    int64 
 6   Picnic     249 non-null    int64 
dtypes: int64(6), object(1)
memory usage: 13.7+ KB


None

In [89]:
print('Nilai yang hilang')
display(df.isnull().sum())

Nilai yang hilang


User Id      0
Sports       0
Religious    0
Nature       0
Theatre      0
Shopping     0
Picnic       0
dtype: int64

In [90]:
# Mengecek outliers dengan menggunakan IQR
numerical_df = df.select_dtypes(include=['int64'])

# Menghitung kuartil pertama (Q1) dan kuartil ketiga (Q3)
Q1 = numerical_df.quantile(0.25)
Q3 = numerical_df.quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Mengecek outliers dengan numerical_data
outliers = ((numerical_df < lower_bound) | (numerical_df > upper_bound)).sum()

print(outliers)


Sports       0
Religious    0
Nature       1
Theatre      3
Shopping     2
Picnic       0
dtype: int64


In [91]:

# Menghapus outliers dari kolom Nature, Theatre, dan Shopping
numerical_df = df[
    (df['Nature'] >= lower_bound['Nature']) & (df['Nature'] <= upper_bound['Nature']) &
    (df['Theatre'] >= lower_bound['Theatre']) & (df['Theatre'] <= upper_bound['Theatre']) &
    (df['Shopping'] >= lower_bound['Shopping']) & (df['Shopping'] <= upper_bound['Shopping'])]

# Menampilkan data setelah penghapusan outliers
print("Data setelah penghapusan outliers:")
numerical_df

Data setelah penghapusan outliers:


User Id  Sports  Religious  Nature  Theatre  Shopping  Picnic
0      User 1       2         77      79       69        68      95
1      User 2       2         62      76       76        69      68
2      User 3       2         50      97       87        50      75
3      User 4       2         68      77       95        76      61
4      User 5       2         98      54       59        95      86
..        ...     ...        ...     ...      ...       ...     ...
244  User 245      18        139     148      129       129     168
245  User 246      22        114     228      104        84     168
246  User 247      20        124     178      104       158     174
247  User 248      20        133     149      139       144     213
248  User 249      20        143     149      139       159     143

[243 rows x 7 columns]

In [92]:
print ('Median')
display(df.median(numeric_only=True))

print ('Mode')
display(df.mode().iloc[0])

print('Standar Deviasi')
display(df.std(numeric_only=True))

print ('Skewness')
display(df.skew(numeric_only=True))

print('Kurtosis')
display(df.kurt(numeric_only=True))


Median


Sports        12.0
Religious    104.0
Nature       119.0
Theatre      113.0
Shopping     104.0
Picnic       119.0
dtype: float64

Mode


User Id      User 1
Sports         12.0
Religious      84.0
Nature         94.0
Theatre       104.0
Shopping       79.0
Picnic        143.0
Name: 0, dtype: object

Standar Deviasi


Sports        6.616501
Religious    32.454115
Nature       45.639372
Theatre      32.132696
Shopping     41.562888
Picnic       32.633339
dtype: float64

Skewness


Sports       0.276508
Religious    0.607930
Nature       0.715522
Theatre      0.579853
Shopping     0.791765
Picnic       0.424635
dtype: float64

Kurtosis


Sports      -1.210310
Religious   -0.131847
Nature       0.536374
Theatre     -0.178279
Shopping     0.010869
Picnic      -0.457196
dtype: float64

In [93]:
numerik = df.corr(numeric_only=True)
numerik

Sports  Religious    Nature   Theatre  Shopping    Picnic
Sports     1.000000   0.623400  0.608372  0.611728  0.583956  0.797777
Religious  0.623400   1.000000 -0.148365  0.169960  0.895652  0.603240
Nature     0.608372  -0.148365  1.000000  0.426901 -0.186125  0.522005
Theatre    0.611728   0.169960  0.426901  1.000000  0.182606  0.225921
Shopping   0.583956   0.895652 -0.186125  0.182606  1.000000  0.442214
Picnic     0.797777   0.603240  0.522005  0.225921  0.442214  1.000000

In [94]:
from sklearn.preprocessing import MinMaxScaler
numerical_columns = ['Sports', 'Religious', 'Nature', 'Theatre', 'Shopping', 'Picnic']

scaler = MinMaxScaler()
df[numerical_columns] = scaler.fit_transform(df[numerical_columns])

merged_df_head =df.head()
merged_df_head


User Id  Sports  Religious    Nature   Theatre  Shopping    Picnic
0  User 1     0.0   0.176471  0.101504  0.064935  0.098361  0.216561
1  User 2     0.0   0.078431  0.090226  0.110390  0.103825  0.044586
2  User 3     0.0   0.000000  0.169173  0.181818  0.000000  0.089172
3  User 4     0.0   0.117647  0.093985  0.233766  0.142077  0.000000
4  User 5     0.0   0.313725  0.007519  0.000000  0.245902  0.159236

In [95]:
from sklearn.preprocessing import MinMaxScaler

columns_to_normalize = ['Sports', 'Religious', 'Nature', 'Theatre', 'Shopping', 'Picnic']

scaler = MinMaxScaler()

df_normalized = df.copy()
df_normalized[columns_to_normalize] = scaler.fit_transform(df[columns_to_normalize])

df_normalized['Total Engagement'] = df[columns_to_normalize].sum(axis=1)
df_normalized['Average Engagement'] = df[columns_to_normalize].mean(axis=1)

df_normalized.head()


User Id  Sports  Religious    Nature   Theatre  Shopping    Picnic  \
0  User 1     0.0   0.176471  0.101504  0.064935  0.098361  0.216561   
1  User 2     0.0   0.078431  0.090226  0.110390  0.103825  0.044586   
2  User 3     0.0   0.000000  0.169173  0.181818  0.000000  0.089172   
3  User 4     0.0   0.117647  0.093985  0.233766  0.142077  0.000000   
4  User 5     0.0   0.313725  0.007519  0.000000  0.245902  0.159236   

   Total Engagement  Average Engagement  
0          0.657831            0.109638  
1          0.427458            0.071243  
2          0.440163            0.073361  
3          0.587475            0.097912  
4          0.726382            0.121064

In [96]:
numerik = df.corr(numeric_only=True)
numerik

Sports  Religious    Nature   Theatre  Shopping    Picnic
Sports     1.000000   0.623400  0.608372  0.611728  0.583956  0.797777
Religious  0.623400   1.000000 -0.148365  0.169960  0.895652  0.603240
Nature     0.608372  -0.148365  1.000000  0.426901 -0.186125  0.522005
Theatre    0.611728   0.169960  0.426901  1.000000  0.182606  0.225921
Shopping   0.583956   0.895652 -0.186125  0.182606  1.000000  0.442214
Picnic     0.797777   0.603240  0.522005  0.225921  0.442214  1.000000

In [97]:
print('Nilai yang hilang')
display(merged_df_head.isnull().sum())

Nilai yang hilang


User Id      0
Sports       0
Religious    0
Nature       0
Theatre      0
Shopping     0
Picnic       0
dtype: int64

In [98]:
from sklearn.decomposition import PCA


pca = PCA(n_components=5) 
reduced_data = pca.fit_transform(df[numerical_columns])


reduced_df = pd.DataFrame(reduced_data, columns=[f'PC{i}' for i in range(1, 6)])


print(reduced_df.head())
print(pca.explained_variance_ratio_)


        PC1       PC2       PC3       PC4       PC5
0 -0.640936  0.104084  0.139490 -0.038273  0.011098
1 -0.736407  0.064482  0.004540  0.067068  0.000002
2 -0.757653 -0.100587  0.016579  0.025430  0.021975
3 -0.690124  0.056480 -0.120916  0.051758 -0.000854
4 -0.585233  0.326068  0.102639  0.002973 -0.026713
[0.61371067 0.23402744 0.11004248 0.02754619 0.0094106 ]


In [99]:
explained_variance = [0.61371067, 0.23402744, 0.11004248, 0.02754619, 0.0094106]

total_variance_explained = sum(explained_variance)

total_variance_explained_percentage = total_variance_explained * 100
total_variance_explained_percentage

99.473738

# Laporan

**Struktur Data**

 Data terdiri dari 249 baris dan 7 kolom. Semua kolom bertipe data numerik kecuali kolom User Id yang bertipe Object. Data tidak memiliki nilai yang hilang (missing value). Selanjutnya, kita akan memeriksa distribusi data dengan menampilkan statistik deskriptif dan histogram untuk setiap kolom minat (kolom selain User Id, menunjukkan tingkat minat pengguna dalam kategori tertentu seperti Olahraga, Keagamaan, Alam, dll.). Kita juga akan menghitung dan menampilkan jumlah outlier pada setiap kolom minat.


**Distribusi Data**

Secara umum, distribusi data untuk setiap kategori minat (interest) cenderung terpusat di sekitar nilai rata-rata, dengan beberapa variasi.
Kolom Nature, Theatre, dan Shopping memiliki outlier, yang menunjukkan adanya beberapa individu dengan minat yang sangat tinggi atau rendah di kategori tersebut.
Tidak ada kolom yang memiliki distribusi yang sangat miring (skewed).



**Penanganan Outlier**

Pada data ini menggunakan teknik Interquartile Range (IQR) untuk mendeteksi outlier. Outlier pada kolom Nature, Theatre, dan Shopping diganti dengan nilai NaN. Setelah itu baris yang mengandung nilai NaN dihapus dari dataset.

**Transformasi Data**

Pada data ini kita menggunakan Teknik min-max scaling, yaitu salah satu metode transformasi data yang digunakan untuk menormalisasi nilai-nilai dalam suatu dataset ke dalam rentang tertentu, biasanya antara 0 dan 1. Metode ini sering digunakan dalam machine learning dan data preprocessing, terutama ketika model atau algoritma sensitif terhadap skala atau rentang nilai data, seperti regresi linear, k-NN, atau neural networks.

**Reduksi Data**

Untuk mereduksi data kita menggunakan metode Principal Component Analysis (PCA), yaitu teknik reduksi dimensi yang banyak digunakan dalam analisis data. PCA bekerja dengan mengubah data berdimensi tinggi menjadi beberapa komponen baru yang saling tidak berkorelasi (principal components), sambil mempertahankan sebanyak mungkin variasi data. PCA sangat berguna dalam menyederhanakan dataset dengan banyak fitur, terutama ketika fitur-fitur tersebut saling berkorelasi.

Persentase total varians yang dijelaskan oleh komponen tersebut adalah 99.47%. Ini berarti bahwa 99.47% dari informasi dalam dataset dipertahankan dalam 5 komponen utama yang dipilih. 